lib/core/src/Cardano/Pool/DB.hs

```diff
@@ -280,6 +288,33 @@ determinePoolLifeCycleStatus mReg mRet = case (mReg, mRet) of
             , show retTime
             ]
 
+-- | Removes all pools with an active retirement epoch that is earlier than
+--   or equal to the specified epoch.
+--
+-- Returns the retirement certificates of the pools that were removed.
+--
+-- See also:
+--
+--    - 'listRetiredPools'.
+--    - 'removePools'.
+--
+removeRetiredPools
+    :: DBLayer IO
+    -> Tracer IO PoolDbLog
+    -> EpochNo
+    -> IO [PoolRetirementCertificate]
+removeRetiredPools
+    DBLayer {atomically, listRetiredPools, removePools} trace epoch =
+        bracketTracer (contramap actionMessage trace) action
+  where
+    actionMessage = MsgRemovingRetiredPoolsForEpoch epoch
+
+    action = atomically (listRetiredPools epoch) >>=
+        \retirementCerts -> do
+            traceWith trace $ MsgRemovingRetiredPools retirementCerts
+            atomically $ removePools (view #poolId <$> retirementCerts)
```

Comment:
The list and remove queries should be within a single `atomically` block. You should be able to liftIO the logging statement.

Reply:
> The list and remove queries should be within a single atomically block. You should be able to liftIO the logging statement.

Good point.

Fixed in 5f9c7df9f9a7c66c38d7bc74ad502737ec12febc.